apps/emqx_machine/src/emqx_machine_terminator.erl

```diff
@@ -0,0 +1,107 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_machine_terminator).
+
+-behaviour(gen_server).
+
+-export([ start/0
+        , graceful/0
+        , graceful_wait/0
+        , is_running/0
+        ]).
+
+-export([init/1, format_status/2,
+         handle_cast/2, handle_call/3, handle_info/2,
+         terminate/2, code_change/3]).
+
+-include_lib("emqx/include/logger.hrl").
+
+-define(TERMINATOR, ?MODULE).
+-define(DO_IT, graceful_shutdown).
+
+%% @doc This API is called to shutdown the Erlang VM by RPC call from remote shell node.
+%% The shutown of apps is delegated to a to a process instead of doing it in the RPC spawned
+%% process which has a remote group leader.
+start() ->
+    {ok, _} = gen_server:start_link({local, ?TERMINATOR}, ?MODULE, [], []),
+    %% NOTE: Do not link this process under any supervision tree
+    ok.
+
+is_running() -> is_pid(whereis(?TERMINATOR)).
+
+%% @doc Send a signal to activate the terminator.
+graceful() ->
+    ?TERMINATOR ! ?DO_IT,
```

Comment:
handle the case where TERMINATOR isn't alive?

Reply:
I think we need some confirmation here, otherwise init:stop/0 will never get called?

Reply:
changed to `gen_server:call`, and `catch` exceptions.